File: packages/domain-types/src/events/workflow.ts
/**
 * Workflow lifecycle domain events (Temporal loop closure — P0).
 *
 * Mirror of `workers/automation/src/jobctrl/domain/events/workflow.py`.
 * Every Temporal workflow durably records a `WorkflowStarted` marker plus
 * exactly one terminal event; together they drive `workflow_run_projections`
 * and the Workflow Runs UI without a trigger-coupled reaper.
 *
 * @see docs/plans/2026-07-03-temporal-native-rearchitecture.md (P0)
 * @see docs/architecture/domain-model/tactical.md §4.7
 */

import type { TenantId } from "../tenant.js";
import { type DomainEvent, createDomainEvent } from "./base.js";

/**
 * The subset of the 12-state `WORKFLOW_RUN_STATUSES` contract that a
 * lifecycle event can carry. Kept as a local union so `@jobctrl/domain-types`
 * stays free of a `@jobctrl/contracts` dependency.
 */
export type WorkflowLifecycleStatus =
  | "in_progress"
  | "succeeded"
  | "failed"
  | "canceled"
  | "timed_out"
  | "terminated";

// -- WorkflowStarted --------------------------------------------------------

export interface WorkflowStartedPayload {
  readonly workflowId: string;
  readonly workflowType: string;
  readonly status: WorkflowLifecycleStatus;
  readonly inputSummary: Record<string, unknown>;
  readonly startedAt: string | null;
  readonly temporalRunId: string | null;
  /**
   * Present only when the reconciler has rediscovered this exact Temporal run
   * after a transient `reconciled_not_found` projection.
   */
  readonly recoveredFromMissingHistory?: boolean;
}

export type WorkflowStarted = DomainEvent<"WorkflowStarted", WorkflowStartedPayload>;

export function createWorkflowStarted(
  tenantId: TenantId,
  payload: WorkflowStartedPayload,
): WorkflowStarted {
  return createDomainEvent("WorkflowStarted", tenantId, payload);
}

// -- WorkflowCancellationRequested -----------------------------------------

export interface WorkflowCancellationRequestedPayload {
  readonly workflowId: string;
  readonly workflowType: string;
  readonly requestedBy: string;
  readonly source: string;
  readonly evidenceKind:
    | "request_intent"
    | "temporal_history"
    | "recovered_temporal_history";
  readonly reason: string | null;
  readonly requestedAt: string;
  readonly temporalRunId: string | null;
  readonly message: string;
}

export type WorkflowCancellationRequested = DomainEvent<
  "WorkflowCancellationRequested",
  WorkflowCancellationRequestedPayload
>;

export function createWorkflowCancellationRequested(
  tenantId: TenantId,
  payload: WorkflowCancellationRequestedPayload,
): WorkflowCancellationRequested {
  return createDomainEvent("WorkflowCancellationRequested", tenantId, payload);
}

// -- WorkflowCompleted ------------------------------------------------------

export interface WorkflowCompletedPayload {
  readonly workflowId: string;
  readonly workflowType: string;
  readonly status: WorkflowLifecycleStatus;
  readonly finishedAt: string | null;
  readonly durationMs: number | null;
  readonly temporalRunId: string | null;
}

export type WorkflowCompleted = DomainEvent<"WorkflowCompleted", WorkflowCompletedPayload>;

export function createWorkflowCompleted(
  tenantId: TenantId,
  payload: WorkflowCompletedPayload,
): WorkflowCompleted {
  return createDomainEvent("WorkflowCompleted", tenantId, payload);
}

// -- WorkflowFailed ---------------------------------------------------------

export interface WorkflowFailedPayload {
  readonly workflowId: string;
  readonly workflowType: string;
  readonly status: WorkflowLifecycleStatus;
  readonly errorCode: string;
  readonly errorMessage: string;
  readonly retryable: boolean;
  readonly finishedAt: string | null;
  readonly durationMs: number | null;
  readonly temporalRunId: string | null;
}

export type WorkflowFailed = DomainEvent<"WorkflowFailed", WorkflowFailedPayload>;

export function createWorkflowFailed(
  tenantId: TenantId,
  payload: WorkflowFailedPayload,
): WorkflowFailed {
  return createDomainEvent("WorkflowFailed", tenantId, payload);
}

// -- WorkflowCanceled -------------------------------------------------------

export interface WorkflowCanceledPayload {
  readonly workflowId: string;
  readonly workflowType: string;
  readonly status: WorkflowLifecycleStatus;
  // Empty on the normal cancel path; the describe-based reconciler fills these
  // with its own provenance when it terminalizes a CANCELED execution.
  readonly errorCode: string;
  readonly errorMessage: string;
  readonly finishedAt: string | null;
  readonly durationMs: number | null;
  readonly temporalRunId: string | null;
}

export type WorkflowCanceled = DomainEvent<"WorkflowCanceled", WorkflowCanceledPayload>;

export function createWorkflowCanceled(
  tenantId: TenantId,
  payload: WorkflowCanceledPayload,
): WorkflowCanceled {
  return createDomainEvent("WorkflowCanceled", tenantId, payload);
}

// -- WorkflowTimedOut -------------------------------------------------------

export interface WorkflowTimedOutPayload {
  readonly workflowId: string;
  readonly workflowType: string;
  readonly status: WorkflowLifecycleStatus;
  readonly errorCode: string;
  readonly errorMessage: string;
  readonly finishedAt: string | null;
  readonly durationMs: number | null;
  readonly temporalRunId: string | null;
}

export type WorkflowTimedOut = DomainEvent<"WorkflowTimedOut", WorkflowTimedOutPayload>;

export function createWorkflowTimedOut(
  tenantId: TenantId,
  payload: WorkflowTimedOutPayload,
): WorkflowTimedOut {
  return createDomainEvent("WorkflowTimedOut", tenantId, payload);
}

// -- WorkflowTerminated -----------------------------------------------------

export interface WorkflowTerminatedPayload {
  readonly workflowId: string;
  readonly workflowType: string;
  readonly status: WorkflowLifecycleStatus;
  readonly errorCode: string;
  readonly errorMessage: string;
  readonly finishedAt: string | null;
  readonly durationMs: number | null;
  readonly temporalRunId: string | null;
}

export type WorkflowTerminated = DomainEvent<"WorkflowTerminated", WorkflowTerminatedPayload>;

export function createWorkflowTerminated(
  tenantId: TenantId,
  payload: WorkflowTerminatedPayload,
): WorkflowTerminated {
  return createDomainEvent("WorkflowTerminated", tenantId, payload);
}

File: workers/automation/src/jobhunter/domain/events/workflow.py
"""Workflow lifecycle domain events (Temporal loop closure — P0).

Every Temporal workflow durably records its own lifecycle: a
``WorkflowStarted`` marker at the top and exactly one terminal event
(``WorkflowCompleted`` / ``WorkflowFailed`` / ``WorkflowCanceled`` /
``WorkflowTimedOut`` / ``WorkflowTerminated``) emitted by the finalize
activity or the describe-based reconciler. Together they make execution
visible in the read-model without a trigger-coupled reaper.

Wire shape: these events carry **camelCase** payload keys because they
cross the JSON-RPC / SSE boundary to the TypeScript read-model, matching
the ``@jobhunter/domain-types`` ``Workflow*`` payload types and the
``GET /v1/events/stream`` contract. The Python fold
(``_rebuild_workflow_runs``) reads the same camelCase keys, so the
finalize writer, the projection reader, and the web handler agree on one
convention.

See ``docs/plans/2026-07-03-temporal-native-rearchitecture.md`` (P0) and
ddd-target.md §4.7 (Pipeline Orchestration).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from jobhunter.domain.tenant import TenantId
from jobhunter.domain.events.base import DomainEvent, create_domain_event

# Terminal status strings live inside the 12-state ``WORKFLOW_RUN_STATUSES``
# contract (packages/contracts/src/schemas.ts). Kept here so the finalize
# activity and the reconciler map to the same vocabulary.
WORKFLOW_STATUS_IN_PROGRESS = "in_progress"
WORKFLOW_STATUS_SUCCEEDED = "succeeded"
WORKFLOW_STATUS_FAILED = "failed"
WORKFLOW_STATUS_CANCELED = "canceled"
WORKFLOW_STATUS_TIMED_OUT = "timed_out"
WORKFLOW_STATUS_TERMINATED = "terminated"


@dataclass(frozen=True)
class WorkflowStartedPayload:
    workflow_id: str
    workflow_type: str
    input_summary: dict[str, Any] = field(default_factory=dict)
    started_at: str | None = None
    temporal_run_id: str | None = None


def create_workflow_started(tenant_id: TenantId, payload: WorkflowStartedPayload) -> DomainEvent:
    return create_domain_event(
        "WorkflowStarted",
        tenant_id,
        {
            "tenantId": str(tenant_id),
            "workflowId": payload.workflow_id,
            "workflowType": payload.workflow_type,
            "status": WORKFLOW_STATUS_IN_PROGRESS,
            "inputSummary": payload.input_summary,
            "startedAt": payload.started_at,
            "temporalRunId": payload.temporal_run_id,
        },
    )


@dataclass(frozen=True)
class WorkflowCompletedPayload:
    workflow_id: str
    workflow_type: str
    finished_at: str | None = None
    duration_ms: int | None = None
    temporal_run_id: str | None = None


def create_workflow_completed(tenant_id: TenantId, payload: WorkflowCompletedPayload) -> DomainEvent:
    return create_domain_event(
        "WorkflowCompleted",
        tenant_id,
        {
            "tenantId": str(tenant_id),
            "workflowId": payload.workflow_id,
            "workflowType": payload.workflow_type,
            "status": WORKFLOW_STATUS_SUCCEEDED,
            "finishedAt": payload.finished_at,
            "durationMs": payload.duration_ms,
            "temporalRunId": payload.temporal_run_id,
        },
    )


@dataclass(frozen=True)
class WorkflowFailedPayload:
    workflow_id: str
    workflow_type: str
    error_code: str = ""
    error_message: str = ""
    retryable: bool = False
    finished_at: str | None = None
    duration_ms: int | None = None
    temporal_run_id: str | None = None


def create_workflow_failed(tenant_id: TenantId, payload: WorkflowFailedPayload) -> DomainEvent:
    return create_domain_event(
        "WorkflowFailed",
        tenant_id,
        {
            "tenantId": str(tenant_id),
            "workflowId": payload.workflow_id,
            "workflowType": payload.workflow_type,
            "status": WORKFLOW_STATUS_FAILED,
            "errorCode": payload.error_code,
            "errorMessage": payload.error_message,
            "retryable": payload.retryable,
            "finishedAt": payload.finished_at,
            "durationMs": payload.duration_ms,
            "temporalRunId": payload.temporal_run_id,
        },
    )


@dataclass(frozen=True)
class WorkflowCanceledPayload:
    workflow_id: str
    workflow_type: str
    # Cancellation carries no app-level failure, so these are empty on the normal
    # path; the describe-based reconciler fills them in with its own provenance
    # (why it closed the row) when it terminalizes a CANCELED execution.
    error_code: str = ""
    error_message: str = ""
    finished_at: str | None = None
    duration_ms: int | None = None
    temporal_run_id: str | None = None


def create_workflow_canceled(tenant_id: TenantId, payload: WorkflowCanceledPayload) -> DomainEvent:
    return create_domain_event(
        "WorkflowCanceled",
        tenant_id,
        {
            "tenantId": str(tenant_id),
            "workflowId": payload.workflow_id,
            "workflowType": payload.workflow_type,
            "status": WORKFLOW_STATUS_CANCELED,
            "errorCode": payload.error_code,
            "errorMessage": payload.error_message,
            "finishedAt": payload.finished_at,
            "durationMs": payload.duration_ms,
            "temporalRunId": payload.temporal_run_id,
        },
    )


@dataclass(frozen=True)
class WorkflowTimedOutPayload:
    workflow_id: str
    workflow_type: str
    error_code: str = ""
    error_message: str = ""
    finished_at: str | None = None
    duration_ms: int | None = None
    temporal_run_id: str | None = None


def create_workflow_timed_out(tenant_id: TenantId, payload: WorkflowTimedOutPayload) -> DomainEvent:
    return create_domain_event(
        "WorkflowTimedOut",
        tenant_id,
        {
            "tenantId": str(tenant_id),
            "workflowId": payload.workflow_id,
            "workflowType": payload.workflow_type,
            "status": WORKFLOW_STATUS_TIMED_OUT,
            "errorCode": payload.error_code,
            "errorMessage": payload.error_message,
            "finishedAt": payload.finished_at,
            "durationMs": payload.duration_ms,
            "temporalRunId": payload.temporal_run_id,
        },
    )


@dataclass(frozen=True)
class WorkflowTerminatedPayload:
    workflow_id: str
    workflow_type: str
    error_code: str = ""
    error_message: str = ""
    finished_at: str | None = None
    duration_ms: int | None = None
    temporal_run_id: str | None = None


def create_workflow_terminated(tenant_id: TenantId, payload: WorkflowTerminatedPayload) -> DomainEvent:
    return create_domain_event(
        "WorkflowTerminated",
        tenant_id,
        {
            "tenantId": str(tenant_id),
            "workflowId": payload.workflow_id,
            "workflowType": payload.workflow_type,
            "status": WORKFLOW_STATUS_TERMINATED,
            "errorCode": payload.error_code,
            "errorMessage": payload.error_message,
            "finishedAt": payload.finished_at,
            "durationMs": payload.duration_ms,
            "temporalRunId": payload.temporal_run_id,
        },
    )
